Sketch of what a prescriptive CodeStyle provider scaffolding could look like

CyrusNajmabadi opened the pull request:
PR for personal review purposes.

Tagging @sharwell .  Thiis is in line with what i was thinkign about based on our gitter convo.  I'm going to document the relevant parts of interest in tihs PR so you can see what i was trying to solve here.

CyrusNajmabadi commented on src/Features/Core/Portable/CodeStyle2/AbstractCodeStyle2Provider.cs:
```diff
@@ -0,0 +1,329 @@
+            Option<CodeStyleOption<TOptionKind>> option,
+            string descriptorId,
+            LocalizableString title,
+            LocalizableString message)
```
subclass would provide this general data.  which would then be used for the analyzer, refactoring, and fix providers.

CyrusNajmabadi commented on src/Features/Core/Portable/CodeStyle2/AbstractCodeStyle2Provider.cs:
```diff
@@ -0,0 +1,329 @@
+        #region diagnostic analyzer
+
+        protected abstract DiagnosticAnalyzerCategory GetDiagnosticAnalyzerCategory();
+        protected abstract bool DiagnosticsForOpenFileOnly(Workspace workspace);
```
subclass still needs to implement these guys.

CyrusNajmabadi commented on src/Features/Core/Portable/CodeStyle2/AbstractCodeStyle2Provider.cs:
```diff
@@ -0,0 +1,308 @@
+            var analyzeCodeBlock = GetCodeBlockAction();
+            if (analyzeCodeBlock != null)
+            {
+                context.RegisterCodeBlockAction(c => AnalyzeIfEnabled(c, analyzeCodeBlock));
```
critically, we want to consolidate the logic about checking if the analyzer should run at all.  i.e. if the user has their option set to 'none' or 'refactoring only' then we do not want the analyzer to run at all.  

To that end, we don't let hte subclass have direct access to the AnalysisContext . Instead, we query them to see what they would like to register.  And then we do the registration for them.  *however*, before we call into their callback, we first do a check if analysis should happen at all.  If not, we never even bother calling into the callback.

CyrusNajmabadi commented on src/Features/Core/Portable/CodeStyle2/AbstractCodeStyle2Provider.cs:
```diff
@@ -0,0 +1,306 @@
+            var (syntaxKinds, analyzeSyntax) = GetSyntaxNodeAction();
+            if (analyzeSyntax != null)
+            {
+                context.RegisterSyntaxNodeAction(c => AnalyzeIfEnabled(c, analyzeSyntax), syntaxKinds);
```
Note: as per above, we control all actual registrations. And we only call the actual callback if the option is at the right severity level.   We'll also do the same when dealing with refactorings.

CyrusNajmabadi commented on src/Features/Core/Portable/CodeStyle2/AbstractCodeStyle2Provider.cs:
```diff
@@ -0,0 +1,306 @@
+            {
+                case ReportDiagnostic.Error:
+                case ReportDiagnostic.Warn:
+                case ReportDiagnostic.Info:
```
@sharwell tehre are 6 values for this enum.  

Default, Error, Warn, Info, Hidden, Suppress.

Error/Warn/Info are all things that should be checked by the analyzer.  

'Hidden' is the same as 'Refactoring' and should *not* be checked by teh analyzer.  

I'm pretty sure 'Suppress' should *not* be checked by the analyzer.

This only leaves 'Default'.  I'm not sure waht that value means and if the analyzer should run when in that state or not.

CyrusNajmabadi commented on src/Features/Core/Portable/CodeStyle2/AbstractCodeStyle2Provider.cs:
```diff
@@ -0,0 +1,306 @@
+                => _codeStyleProvider.GetDiagnosticAnalyzerCategory();
+
+            public sealed override bool OpenFileOnly(Workspace workspace)
+                => _codeStyleProvider.DiagnosticsForOpenFileOnly(workspace);
```
all members in these classes are intentionally sealed.  All logic should be provided in the CodeStyleProvider subclass.

CyrusNajmabadi commented on src/Features/Core/Portable/CodeStyle2/AbstractCodeStyle2Provider.cs:
```diff
@@ -0,0 +1,306 @@
+            return false;
+        }
+
+        public abstract class CodeStyle2DiagnosticAnalyzer : AbstractCodeStyleDiagnosticAnalyzer
```
here are the impls of hte actual DiagnosticAnalyzer/CodeRefactoringProvider/CodeFixProvider.  They just defer everything to the AbstractCodeStyleProvider.  Ideally they would be sealed types.  However, because people have to export DiagnosticAnalyzer/CodeRefactoringProvider/CodeFixProvider themselves, these are exposed and abstract *solely* for the purpose of creating an exported subclass.  You'll see that at the bottom of this file.

CyrusNajmabadi commented on src/Features/Core/Portable/CodeStyle2/AbstractCodeStyle2Provider.cs:
```diff
@@ -0,0 +1,311 @@
+                    // the feature here as a refactoring.
+                    await ComputeForwardRefactoringsAsync(context).ConfigureAwait(false);
+                    break;
+            }
```
@sharwell as above, i'm not sure what the right behavior is for ReportDiagnotic.Default.  Can you help me understand what that value means?

CyrusNajmabadi commented on src/Features/Core/Portable/CodeStyle2/AbstractCodeStyle2Provider.cs:
```diff
@@ -0,0 +1,312 @@
+                    // we shouldn't really be blocking the user from making it.
+                    await ComputeReverseRefactoringsAsync(context).ConfigureAwait(false);
+                    break;
+            }
```
@sharwell as above, i'm not sure what behavior to have for ReportDiagnostic.Default.  Can you help explain to me what that value means?

CyrusNajmabadi commented on src/Features/Core/Portable/CodeStyle2/AbstractCodeStyle2Provider.cs:
```diff
@@ -0,0 +1,331 @@
+    [ExportCodeFixProvider(LanguageNames.CSharp, Name = "Example code fix provider")]
+    class DummyCodeFixProvider : DummyCodeStyleProvider.CodeStyle2CodeFixProvider
+    {
+    }
```
These are the three classes a feature would still have to export.  Note they are very intentionally empty.  They do not need to do anything, and literally exist *solely* so we can add the export attributes somewhere so these can be found by the relevant subsystems.

in an ideal world we could just put these attributes on the CodeStyleProvider.  however, since each of these attributes expects a particular inheritance hierarchy, and because they all use classes, there is no way to make CodeStyleProvider extend all three of these classes.

CyrusNajmabadi replied:
Alternatively, we could just have something like ExportCodeStyleProvider, and have the relevant subsystems know how to import that.  Then, the CodeStyleProvider would expose the 3 separate providers from itself.  That would be trivial to do for CodeRefactoringProvider/CodeFixProvider.  however, it's unclear to me how we would fit that into the DiagnosticAnalyzer infrastructure.

CyrusNajmabadi commented on src/Features/Core/Portable/CodeStyle2/AbstractCodeStyle2Provider.cs:
```diff
@@ -0,0 +1,331 @@
+    {
+    }
+
+    class DummyCodeStyleProvider : AbstractCodeStyle2Provider<ExpressionBodyPreference, int, DummyCodeStyleProvider>
```
an example of what a code style provider would look like.

CyrusNajmabadi commented on src/Features/Core/Portable/CodeStyle2/AbstractCodeStyle2Provider.cs:
```diff
@@ -0,0 +1,339 @@
+        private void AnalyzeSyntax(SyntaxNodeAnalysisContext obj)
+        {
+            throw new NotImplementedException();
+        }
```
subclass would implement this.  but it would only actually be called if appropriate.  subclass would not have to actually have to have the logic about "should i be runnign... or should only teh refactoring be running?" :)

CyrusNajmabadi commented on src/Features/Core/Portable/CodeStyle2/AbstractCodeStyle2Provider.cs:
```diff
@@ -0,0 +1,352 @@
+            analyze(context, optionValue.Value);
+        }
+
+        public abstract class CodeStyle2DiagnosticAnalyzer : AbstractCodeStyleDiagnosticAnalyzer
```
here are the impls of hte actual DiagnosticAnalyzer/CodeRefactoringProvider/CodeFixProvider. They just defer everything to the AbstractCodeStyleProvider. Ideally they would be sealed types. However, because people have to export DiagnosticAnalyzer/CodeRefactoringProvider/CodeFixProvider themselves, these are exposed and abstract solely for the purpose of creating an exported subclass. You'll see that at the bottom of this file.

CyrusNajmabadi commented on src/Features/Core/Portable/CodeStyle2/AbstractCodeStyle2Provider.cs:
```diff
@@ -0,0 +1,352 @@
+                    // just cause an error.  That said, maybe this is just an intermediary step, and
+                    // we shouldn't really be blocking the user from making it.
+                    await ComputeRefactoringsAsync(context, optionValue.Value, forward: false).ConfigureAwait(false);
+                    return;
```
@sharwell as above, i'm not sure what the right behavior is for ReportDiagnotic.Default. Can you help me understand what that value means?

CyrusNajmabadi commented on src/Features/Core/Portable/CodeStyle2/AbstractCodeStyle2Provider.cs:
```diff
@@ -0,0 +1,352 @@
+                    // that we shouldn't even offer to refactor in the reverse direction if it will
+                    // just cause an error.  That said, maybe this is just an intermediary step, and
+                    // we shouldn't really be blocking the user from making it.
+                    await ComputeRefactoringsAsync(context, optionValue.Value, forward: false).ConfigureAwait(false);
```
this is one of the interesting parts about analysis/refactoring that we can code in as a general pattern. Specifically, if someone has a code style, and has it set to info/warn/error. Then the analyzer will in general find code that is out of spec and will tell the user to fix it. However, for code that is following spec, we want to offer the "reverse refactoring" to allow them a quick way to change teh style to the non-preferred form.

CyrusNajmabadi commented on src/Features/Core/Portable/CodeStyle2/AbstractCodeStyle2Provider.cs:
```diff
@@ -0,0 +1,352 @@
+        protected abstract Task<ImmutableArray<CodeAction>> ComputeForwardRefactoringsAsync(
+            Document document, TextSpan span, TOptionKind option, CancellationToken cancellationToken);
+        protected abstract Task<ImmutableArray<CodeAction>> ComputeReverseRefactoringsAsync(
+            Document document, TextSpan span, TOptionKind option, CancellationToken cancellationToken);
```
this forces the subclass to realize that it needs to both offer the normal refactoring, as well as the reverse refactoring.

CyrusNajmabadi commented on src/Features/Core/Portable/CodeStyle2/AbstractCodeStyle2Provider.cs:
```diff
@@ -0,0 +1,352 @@
+    [ExportCodeFixProvider(LanguageNames.CSharp, Name = "Example code fix provider")]
+    class ExampleCodeFixProvider : ExampleCodeStyleProvider.CodeStyle2CodeFixProvider
+    {
+    }
```
These are the three classes a feature would still have to export. Note they are very intentionally empty. They do not need to do anything, and literally exist solely so we can add the export attributes somewhere so these can be found by the relevant subsystems.

in an ideal world we could just put these attributes on the CodeStyleProvider. however, since each of these attributes expects a particular inheritance hierarchy, and because they all use classes, there is no way to make CodeStyleProvider extend all three of these classes.

Alternatively, we could just have something like ExportCodeStyleProvider, and have the relevant subsystems know how to import that. Then, the CodeStyleProvider would expose the 3 separate providers from itself. That would be trivial to do for CodeRefactoringProvider/CodeFixProvider. however, it's unclear to me how we would fit that into the DiagnosticAnalyzer infrastructure.

CyrusNajmabadi commented on src/Features/Core/Portable/CodeStyle2/AbstractCodeStyle2Provider.cs:
```diff
@@ -0,0 +1,352 @@
+        private void AnalyzeSyntax(SyntaxNodeAnalysisContext context, ExpressionBodyPreference preference)
+        {
+            throw new NotImplementedException();
+        }
```
subclass would implement this. but it would only actually be called if appropriate. subclass would not have to actually have to have the logic about "should i be runnign... or should only teh refactoring be running?" :)

CyrusNajmabadi commented on src/Features/Core/Portable/CodeStyle2/AbstractCodeStyle2Provider.Analysis.cs:
```diff
@@ -0,0 +1,90 @@
+                    break;
+                default:
+                    // don't analyze if it's any other value.
+                    return;
```
@sharwell tehre are 6 values for this enum.

Default, Error, Warn, Info, Hidden, Suppress.

Error/Warn/Info are all things that should be checked by the analyzer.

'Hidden' is the same as 'Refactoring' and should not be checked by teh analyzer.

I'm pretty sure 'Suppress' should not be checked by the analyzer.

This only leaves 'Default'. I'm not sure waht that value means and if the analyzer should run when in that state or not.

CyrusNajmabadi commented on src/Features/Core/Portable/CodeStyle2/AbstractCodeStyle2Provider.Analysis.cs:
```diff
@@ -0,0 +1,90 @@
+                => _codeStyleProvider.GetDiagnosticAnalyzerCategory();
+
+            public sealed override bool OpenFileOnly(Workspace workspace)
+                => _codeStyleProvider.DiagnosticsForOpenFileOnly(workspace);
```
all members in these classes are intentionally sealed. All logic should be provided in the CodeStyleProvider subclass.

CyrusNajmabadi commented on src/Features/Core/Portable/CodeStyle2/AbstractCodeStyle2Provider.Analysis.cs:
```diff
@@ -0,0 +1,90 @@
+            analyze(context, optionValue.Value);
+        }
+
+        public abstract class DiagnosticAnalyzer : AbstractCodeStyleDiagnosticAnalyzer
```
here are the impls of hte actual DiagnosticAnalyzer/CodeRefactoringProvider/CodeFixProvider. They just defer everything to the AbstractCodeStyleProvider. Ideally they would be sealed types. However, because people have to export DiagnosticAnalyzer/CodeRefactoringProvider/CodeFixProvider themselves, these are exposed and abstract solely for the purpose of creating an exported subclass. You'll see that at the bottom of this file.

CyrusNajmabadi commented on src/Features/Core/Portable/CodeStyle2/AbstractCodeStyle2Provider.Refactoring.cs:
```diff
@@ -0,0 +1,89 @@
+                    // we shouldn't really be blocking the user from making it.
+                    await ComputeRefactoringsAsync(context, optionValue.Value, forward: false).ConfigureAwait(false);
+                    return;
+            }
```
@sharwell as above, i'm not sure what the right behavior is for ReportDiagnotic.Default. Can you help me understand what that value means?

CyrusNajmabadi commented on src/Features/Core/Portable/CodeStyle2/AbstractCodeStyle2Provider.Refactoring.cs:
```diff
@@ -0,0 +1,89 @@
+                    // that we shouldn't even offer to refactor in the reverse direction if it will
+                    // just cause an error.  That said, maybe this is just an intermediary step, and
+                    // we shouldn't really be blocking the user from making it.
+                    await ComputeRefactoringsAsync(context, optionValue.Value, forward: false).ConfigureAwait(false);
```
this is one of the interesting parts about analysis/refactoring that we can code in as a general pattern. Specifically, if someone has a code style, and has it set to info/warn/error. Then the analyzer will in general find code that is out of spec and will tell the user to fix it. However, for code that is following spec, we want to offer the "reverse refactoring" to allow them a quick way to change teh style to the non-preferred form.

CyrusNajmabadi commented on src/Features/Core/Portable/CodeStyle2/ExampleCodeStyleProvider.cs:
```diff
@@ -0,0 +1,106 @@
+    [ExportCodeFixProvider(LanguageNames.CSharp, Name = "Example code fix provider")]
+    class ExampleCodeFixProvider : ExampleCodeStyleProvider.CodeFixProvider
+    {
+    }
```
These are the three classes a feature would still have to export. Note they are very intentionally empty. They do not need to do anything, and literally exist solely so we can add the export attributes somewhere so these can be found by the relevant subsystems.

in an ideal world we could just put these attributes on the CodeStyleProvider. however, since each of these attributes expects a particular inheritance hierarchy, and because they all use classes, there is no way to make CodeStyleProvider extend all three of these classes.

Alternatively, we could just have something like ExportCodeStyleProvider, and have the relevant subsystems know how to import that. Then, the CodeStyleProvider would expose the 3 separate providers from itself. That would be trivial to do for CodeRefactoringProvider/CodeFixProvider. however, it's unclear to me how we would fit that into the DiagnosticAnalyzer infrastructure.

CyrusNajmabadi commented on src/Features/Core/Portable/CodeStyle2/ExampleCodeStyleProvider.cs:
```diff
@@ -0,0 +1,106 @@
+        private void AnalyzeSyntax(SyntaxNodeAnalysisContext context, ExpressionBodyPreference preference)
+        {
+            throw new NotImplementedException();
+        }
```
subclass would implement this. but it would only actually be called if appropriate. subclass would not have to actually have to have the logic about "should i be running... or should only the refactoring be running?" :)

CyrusNajmabadi commented:
Tagging @mavasani as well.  The intent here is to provide a suitable base-class that codifies our general patterns around codestyle refactorings and fixers.  With this, a feature (like ```use-expression-body```, or ```use-this.```) would have a simple class they could subclass that would then prescriptively guide their implementation. 

Specifically, it would force them to have the appropriate analysis code and refactoring code, without a lot of the boilerplate options-checking/switching code necessary.  Trying to get this logic right is actually fairly tricky, and many features have had bugs in this area.

For example, many features are inconsistent about checking things like "is this feature on or off?"  This scaffolding actually handles all of that, so the feature doesn't have to care.  The relevant derived type methods will only be called when actually relevant if the feature is actually 'on'.

We've also had the problem of many code-style features not actually being consistent in terms of the experience we want.  For example, even when user code *fits* the specified code-style, we generally want to expose a refactoring to let the user flip to the *opposite* code style.  i.e. using an explicit type, even if 'var' is preferred (or vice versa).  Some features do this.  But it's been haphazardly done.

With this scaffolding, you are forced to provide this functionality.  The feature impl *must* support:

1. analyzing code to make sure it it compliant (when the user config is info/warning/error).
2. refactoring code to make it compliant (when the user config is 'refactoring-only')
3. a reverse-refactoring to allow users to take compliant code and switch to the non-compliant style.

CyrusNajmabadi commented:
@mavasani @sharwell I have this working with a real codestyle feature here: https://github.com/dotnet/roslyn/pull/28770

The relevant file of that PR is https://github.com/dotnet/roslyn/pull/28770/files#diff-5d64fb2ae069850f4d5be1cdea5c6f46

There you can see an entire complex code-style feature that will work properly as a code-fixer or refactoring.

CyrusNajmabadi commented on src/Features/Core/Portable/CodeStyle/AbstractCodeStyleProvider.cs:
```diff
@@ -0,0 +1,134 @@
+            _option = option;
+            _descriptorId = descriptorId;
+            _title = title;
+            _message = message;
```
subclass would provide this general data. which would then be used for the analyzer, refactoring, and fix providers.

CyrusNajmabadi commented on src/Features/Core/Portable/CodeStyle/AbstractCodeStyleProvider.Analysis.cs:
```diff
@@ -0,0 +1,98 @@
+            if (analyzeSemanticModel != null) { context.RegisterSemanticModelAction(c => AnalyzeIfEnabled(c, analyzeSemanticModel)); }
+            if (analyzeSyntaxTree != null) { context.RegisterSyntaxTreeAction(c => AnalyzeIfEnabled(c, analyzeSyntaxTree)); }
+            if (analyzeSyntax != null) { context.RegisterSyntaxNodeAction(c => AnalyzeIfEnabled(c, analyzeSyntax), syntaxKinds); }
+            if (analyzeOperation != null) { context.RegisterOperationAction(c => AnalyzeIfEnabled(c, analyzeOperation), operationKinds); }
```
critically, we want to consolidate the logic about checking if the analyzer should run at all. i.e. if the user has their option set to 'none' or 'refactoring only' then we do not want the analyzer to run at all.

To that end, we don't let hte subclass have direct access to the AnalysisContext . Instead, we query them to see what they would like to register. And then we do the registration for them. however, before we call into their callback, we first do a check if analysis should happen at all. If not, we never even bother calling into the callback.

CyrusNajmabadi commented on src/Features/Core/Portable/CodeStyle/AbstractCodeStyleProvider.Analysis.cs:
```diff
@@ -0,0 +1,98 @@
+                default:
+                    // don't analyze if it's any other value.
+                    return;
+            }
```
@sharwell tehre are 6 values for this enum.

Default, Error, Warn, Info, Hidden, Suppress.

Error/Warn/Info are all things that should be checked by the analyzer.

'Hidden' is the same as 'Refactoring' and should not be checked by teh analyzer.

I'm pretty sure 'Suppress' should not be checked by the analyzer.

This only leaves 'Default'. I'm not sure waht that value means and if the analyzer should run when in that state or not.

CyrusNajmabadi commented on src/Features/Core/Portable/CodeStyle/AbstractCodeStyleProvider.Analysis.cs:
```diff
@@ -0,0 +1,98 @@
+                => _codeStyleProvider.GetDiagnosticAnalyzerCategory();
+
+            public sealed override bool OpenFileOnly(Workspace workspace)
+                => _codeStyleProvider.DiagnosticsForOpenFileOnly(workspace);
```
all members in these nested classes are intentionally sealed. All logic should be provided in the CodeStyleProvider subclass.  The feature classes that subclass these nested classes should nto be able to override these themselves.

CyrusNajmabadi commented on src/Features/Core/Portable/CodeStyle/AbstractCodeStyleProvider.Analysis.cs:
```diff
@@ -0,0 +1,98 @@
+            analyze(context, optionValue);
+        }
+
+        public abstract class DiagnosticAnalyzer : AbstractCodeStyleDiagnosticAnalyzer
```
here are the impls of hte actual DiagnosticAnalyzer/CodeRefactoringProvider/CodeFixProvider. They just defer everything to the AbstractCodeStyleProvider. Ideally they would be sealed types. 

However, because people have to export DiagnosticAnalyzer/CodeRefactoringProvider/CodeFixProvider themselves, these are exposed and abstract solely for the purpose of creating an exported subclass. You'll see that later on.

CyrusNajmabadi commented on src/Features/Core/Portable/CodeStyle/AbstractCodeStyleProvider.Refactoring.cs:
```diff
@@ -0,0 +1,89 @@
+                    // we shouldn't really be blocking the user from making it.
+                    await ComputeRefactoringsAsync(context, optionValue.Value, forward: false).ConfigureAwait(false);
+                    return;
+            }
```
@sharwell as above, i'm not sure what the right behavior is for ReportDiagnotic.Default. Can you help me understand what that value means?

CyrusNajmabadi commented on src/Features/Core/Portable/CodeStyle/AbstractCodeStyleProvider.Refactoring.cs:
```diff
@@ -0,0 +1,89 @@
+                    // just cause an error.  That said, maybe this is just an intermediary step, and
+                    // we shouldn't really be blocking the user from making it.
+                    await ComputeRefactoringsAsync(context, optionValue.Value, forward: false).ConfigureAwait(false);
+                    return;
```
this is one of the interesting parts about analysis/refactoring that we can code in as a general pattern. Specifically, if someone has a code style, and has it set to info/warn/error. Then the analyzer will in general find code that is out of spec and will tell the user to fix it. However, for code that is following spec, we want to offer the "reverse refactoring" to allow them a quick way to change teh style to the non-preferred form.

CyrusNajmabadi commented on src/Features/Core/Portable/CodeStyle/AbstractCodeStyleProvider.cs:
```diff
@@ -0,0 +1,134 @@
+        /// expression body already.
+        /// </summary>
+        protected abstract Task<ImmutableArray<CodeAction>> ComputeReverseRefactoringsAsync(
+            Document document, TextSpan span, TOptionKind option, CancellationToken cancellationToken);
```
Note: i welcome different names than 'Forward' and 'Reverse' refactoring.  If anyone has any good ideas here, let me know!

CyrusNajmabadi replied:
Perhaps:

```ComputeRefactoringsWhenAnalyzerNotPresent``` and
```ComputeOppositeRefactoringsWhenAnalyzerPresent```

?

Neme12 commented:
@CyrusNajmabadi I remember reading your comment about whether we should offer a refactoring against the user's preference even if its severity is an error (but can't find the comment now). My first thought was that it seems reasonable to not offer a refactoring if it would introduce an error or a warning.

But consider this scenario: the user prefers expression bodied methods and it is enforced with an error. But now the user wants to add another statement to the method and in order to do that, it has to be converted to a block body as an intermediate step.

CyrusNajmabadi commented:
> But consider this scenario: the user prefers expression bodied methods and it is enforced with an error. But now the user wants to add another statement to the method and in order to do that, it has to be converted to a block body as an intermediate step.

Yup.  Agreed.

CyrusNajmabadi commented:
Having used this for "use expression bodies for lambdas" i *really* like this scaffolding.  It made the feature very easy to write in the *correct* way.

CyrusNajmabadi commented:
@mavasani Can you take a look and tell me what you think?  For reference, here's an example of a feature that demonstrates using this new scaffolding: https://github.com/dotnet/roslyn/pull/28770

I think it works quite nicely and helps ensure that:

1. implementations do not have to repeat common logic about if/when analyzers vs refactorings should run.
2. implementations can share code easily across their analyzer, refactoring and fixers.
3. implementations know what pattern they have to provide, especially around the interaction of their analyzer and refactoring portion.

CyrusNajmabadi commented:
@mavasani @sharwell @jcouv @dotnet/roslyn-ide 

Can i get eyes on this?  I'm very curious how people feel about this sort of unified prescriptive scaffolding approach to analyzers/refactorings/fixers.

I personally think its good given that it forces implementations to have to consider many aspects of how all these interact.  Currently, many (most?) of our features get this wrong.  And getting it right can be quite tricky.  This class takes away most of that complexity, allowing the implementation to just focus on the feature.

CyrusNajmabadi commented on src/Features/Core/Portable/CodeStyle/AbstractCodeStyleProvider.Analysis.cs:
```diff
@@ -0,0 +1,139 @@
+    // subclasses cannot change anything.  All code relevant to subclasses relating to analysis
+    // is contained in AbstractCodeStyleProvider.cs
+
+    internal abstract partial class AbstractCodeStyleProvider<TOptionKind, TSyntaxKind, TCodeStyleProvider>
```
No.  THey can be language agnostics.  But if they are, they'll have to use 'int' here (the same kind as RawKind).

CyrusNajmabadi commented on src/Features/Core/Portable/CodeStyle/AbstractCodeStyleProvider.Analysis.cs:
```diff
@@ -0,0 +1,139 @@
+        }
+
+        /// <summary>
+        /// Critically, we want to consolidate the logic about checking if the analyzer should run
```
Ok.  So where does that get done?  And how is it done.  If someone doesn't know about options, how do they know what sort of diagnostic to produce (since it is option-value dependent)?

sharwell replied:
The implementation here is a unification of behavior we have today. I'm interested in alternatives such as #27712 but I see no reason why that open design work would block forward progress of this as a clarifying/supporting change.

CyrusNajmabadi commented on src/Features/Core/Portable/CodeStyle/AbstractCodeStyleProvider.Analysis.cs:
```diff
@@ -0,0 +1,139 @@
+                TCodeStyleProvider provider, TContext context, Action<TContext, CodeStyleOption<TOptionKind>> analyze,
+                AnalyzerOptions options, SyntaxTree syntaxTree, CancellationToken cancellationToken)
+            {
+                var optionSet = options.GetDocumentOptionSetAsync(syntaxTree, cancellationToken).GetAwaiter().GetResult();
```
Yes.  I would be fine with the engine computing this.  Let me know when that is done and we can remove some of this code :)

sharwell replied:
This implementation aligns with what analyzers do today, so it's not inherently a problem. I wouldn't spend too much time changing it to something else considering it will be replaced by the new .editorconfig-in-compiler functionality.

CyrusNajmabadi commented on src/Features/Core/Portable/CodeStyle/AbstractCodeStyleProvider.Refactoring.cs:
```diff
@@ -0,0 +1,89 @@
+            var optionSet = await document.GetOptionsAsync(cancellationToken).ConfigureAwait(false);
+            var optionValue = optionSet.GetOption(_option);
+
+            var severity = GetOptionSeverity(optionValue);
```
> and/or we can consider adding a new CodeFixAndRefactoringProvider API for such providers. The engine can then do the right thing based on user settings. One of the biggest benefits here would be that third party providers can also benefit from it.

Well, that's what this work is trying to be the first step toward :)   The idea is not: "This is what we would ship to customers".  The idea is: "here are the parts we would control, and here are the extension points a feature would have to implement".

CyrusNajmabadi commented:
Note: at the very least *in the interim* period, this would serve as a good system for existing and new *IDE* features to use.  IT would help us shake out so many bugs where existing features behave inconsistently (especially around the concept of computing the 'Opposing Refactoring').  

Then, if most/all of the IDE was on this system, it would be much easier to move to whatever new system the analysis engine ends up producing.

What do you think @mavasani ?

mavasani commented:
> Then, if most/all of the IDE was on this system, it would be much easier to move to whatever new system the analysis engine ends up producing.

Sounds good to me, though I will let @sharwell make the final call here as he is driving quite a few things that are related (move CodeStyle providers to NuGet, complimentary fixer/refactoring lightup, etc.).

CyrusNajmabadi commented:
> Sounds good to me, though I will let @sharwell make the final call here as he is driving quite a few things that are related (move CodeStyle providers to NuGet, complimentary fixer/refactoring lightup, etc.).

That sounds reasonable to me.  Fortunately, i think a lot of these goals are complementary.  My primary goal (which it seems you are interested in as well) is to try to get individual analzyers/fixers out of hte job of having to understand all these subtleties about options and behavior.  They should have a prescriptive mechanism that effectively says: just implements X, Y and Z, and things shoudl work properly.  

I would def love that in the engine itself.  However, in the interim period, i'm happy to try to codify within the IDE itself at some non-extendible point.

CyrusNajmabadi commented:
@sharwell @jinujoseph  Can you look please.  THis PR is a month old now, and i would like to have it merged so that i can use it for the Lambda PR (https://github.com/dotnet/roslyn/pull/28765) (also waiting on you).  :)   The demonstration on how this cleans things up can be found here: https://github.com/dotnet/roslyn/pull/28770

CyrusNajmabadi commented:
Actually, @jcouv Please look at thsi PR instead of https://github.com/dotnet/roslyn/pull/28770.  it will help keep the new system clear, before you have to see it in action with an existing feature.

CyrusNajmabadi commented on src/Features/Core/Portable/CodeStyle/AbstractCodeStyleProvider.cs:
```diff
@@ -0,0 +1,141 @@
+            var severity = optionValue.Notification.Severity;
+            return severity == ReportDiagnostic.Default
+                ? severity.WithDefaultSeverity(DiagnosticSeverity.Hidden)
+                : severity;
```
@sharwell said this is the right thing to do with ReportDiagnostic is set to Default.

CyrusNajmabadi replied:
gotcha.  I thought from a previous discussion you recommended the above in case that behavior would change.  But i'm happy to simplify if you think it's the right thing to do.

CyrusNajmabadi commented:
@mavasani Can you sign off if you're ok with this, or let me know what you'd like changed?  

Note: i do view this as a nice intermediary stepping stone that is complimentary with your efforts and with Sam's efforts.  For example, if/when we can just make the engine understand options, we can remove the code in here that checks that against each document.  However, in the interim period, this is still a good thing to have as it is something every IDE feature needs to do, and this codifies it in one place.  The more features i can move over to this system, the *easier* it will be to then move them all at once onto whatever you build.

Thanks!

mavasani commented on src/Features/Core/Portable/CodeStyle/AbstractCodeStyleProvider.Fixing.cs:
```diff
@@ -0,0 +1,48 @@
+            protected CodeFixProvider()
+            {
+                _codeStyleProvider = new TCodeStyleProvider();
+                FixableDiagnosticIds = ImmutableArray.Create(_codeStyleProvider._descriptorId);
```
How about fixers that can fix multiple diagnostic IDs?

CyrusNajmabadi replied:
You're right. We'll probably need to handle that at some point.  I could do that as part of this PR, or wait until we need it.  I'd prefer hte latter :)

CyrusNajmabadi commented:
@mavasani Plz merge when convenient.  Thanks!

mavasani commented:
> @mavasani Plz merge when convenient. Thanks!

I still would prefer if @sharwell reviews this PR and decides if/when this is good to merge.

CyrusNajmabadi commented:
@sharwell can you please take a look.  At the very least, this serves as a good harness in which to move existing IDE codestyle features to.  Taht can help bring consistency between the features (especially around Fixes/Refactorings/Options), which will then make it easier to move to anything else in the future.

heejaechang commented on src/Features/Core/Portable/CodeStyle/AbstractCodeStyleProvider.Analysis.cs:
```diff
@@ -0,0 +1,139 @@
+                }
+
+                var optionValue = optionSet.GetOption(provider._option);
+                var severity = GetOptionSeverity(optionValue);
```
how does this option severity thing work with compiler severity? it seems to get options from GetDocumentOptionAsync, what happen if user set options through rule set?

we let IDE team to do this as a hack before. are we keeping this hack? whether something run or not should be engines decision not each analyzer.

if IDE wants to create its own system, it should build its own system rather than breaking contracts/assumptions on the existing system created for compiler's diagnostic analyzers.

heejaechang replied:
I believe the ultimately, severity is not something analyzer author will use since they can't know what is the last effective severity will be that went through all error mapping such as rule set, warnaserror and etc.

but it is the engine which decide what to run and when based one severity such as hidden, none or refactoring only.

CyrusNajmabadi replied:
> we let IDE team to do this as a hack before. 

I don't know what this means.  IDE didn't do anything here.  This is the API the analyzer system provided to support the options system the IDE has had since some of hte earliest days of Roslyn.  If you want another system used, can you please provide such a system?  IDE is simply using what provided and recommended that it use *by* the analyzer team.

> if IDE wants to create its own system, it should build its own system rather than breaking contracts/assumptions on the existing system created for compiler's diagnostic analyzers.

The diagnostic analyzer system was built *from day one* to be a system that needed to be suitable for IDE scenarios.  The APIs (including teh options system) was provided by the analyzer system.  The IDE never created its own system.  It used what was provided.  Note: the analyzer subsytem was provided by teh analyzer team *and* many features were ported over to use it by that team.  If this system is not sufficient, then there's nothing the IDE can do about that.  We depend on the APIs supplied by the analyzer component.

> but it is the engine which decide what to run and when based one severity such as hidden, none or refactoring only.

Great.  The moment the engine can provide that information, we can consume it on the IDE side.  Until then, i have no idea what hte IDE should do since this is the *only* API available to do this computation.

heejaechang replied:
severity is not something analyzer should be able to play with. but IDE team wanted it since they didn't want to use ruleset editor already provided but provide option page in IDE which we have been telling third part analyzer not to do.

so that is what I meant by hack.

...

sam is working on the right way to handle this. which takes time. 

..

> The diagnostic analyzer system was built from day one to be a system that needed to be suitable for IDE scenarios. The APIs (including teh options system) was provided by the analyzer system.

no. diagnostic analyzer is built from day one for compiler. we had some thing for IDE which we throw away for new diagnostic analyzer built for compiler. that's what I mean by creating IDE own stuff if diagnostic analyzer built for compiler is not flexible enough to support everything IDE wants.

since compiler doesn't work the way IDE work, it let something that is hard for IDE to support such as compiler end analyzer in good perf and also, in other cases, it doesn't support things IDE want to support such as solution wide analyzer.

to let system built for compiler to work the way IDE wants, we introduced hacks including analyzer itself checking severity and change its behavior.

heejaechang commented:
having helper to make some analyzer/fixer easier to write seems fine. but if it again becomes something forcing other to all use like tagger, I would rather not accept this.

CyrusNajmabadi commented:
> having helper to make some analyzer/fixer easier to write seems fine. but if it again becomes something forcing other to all use like tagger,

I don't know what that means.  There is nothing aobut the tagging infrastructure that forces anything.  People are always free to use ITagger or IAccurateTagger.  The IDE does have scaffolding for async tagging.  And this is a *good* thing.  It keeps nearly every IDE tagger from having to reimplement the same logic which they would likely not get right.  It means we can consolidate patterns and behaviors in one location to provide a consistent experience and easy to implement surface area.  Furthermore, when we want to fix something we can do it in one place.

> I would rather not accept this.  

Do you have a recommendation on what a better course of action is?  Right now, it's safe to say that nearly *all* IDE refactorings/fixes get this wrong.  They don't properly behave in a consistent manner, and it's very annoying and difficult to get this logic right.  Providing a base implementation means we can remove complex and unnecessary code from each analyzers/fixer/refactoring, and can unify it in one place.

heejaechang commented:
> I don't know what that means. There is nothing aobut the tagging infrastructure that forces anything. People are always free to use ITagger or IAccurateTagger. 

not sure what to say. I wrote diagnostc tagger not using the system and I can show you all the comment you wrote on I should be using the tagger infrastructure. and then once used, it can't do anything the infrastructure is not built for and hard to add new ability since things are so tightly coupled.

I don't want to go into argument. it people are fine to use other things, that's fine then. if they feel it is worth, then they will use. if they feel it is more work to use it then won't.

...

> Right now, it's safe to say that nearly all IDE refactorings/fixes get this wrong. They don't properly behave in a consistent manner

if it really can have simple base case for all analyzer, sure fine. but if it has to have all this built-in knob for one-off different behavior like tagger then not sure. and later things become so hard to add new thing and must bend what one wants to do since the helper is not built to do such a thing.

...

> Do you have a recommendation on what a better course of action is?

I would prefer bunch of library/helper that one can reuse. there can be some base type like this but those bunch of library/helper will let people to create different type but share behavior for one off behavior. rather than 1 base type having bunch of knobs for each one-off behavior.

CyrusNajmabadi commented:
>no. diagnostic analyzer is built from day one for compiler.

That is 100% incorrect.  I was in the meetings on this from day 1.  A core goal from the very first meeting was that this system needed to be sufficient for IDE needs.  This was non negotiable and it was hte only reason IDE even allowed all our features to be moved onto it in the first place.  If the analyzer system could not meet those needs, we would not have moved things over to it.

From the beginning of this the IDE has only used APIs that were delivered by the analyzer team.  At no point has the IDE attempted to 'hack' around anything.  We've only used the APIs that were provided in the manner we were told was correct.

CyrusNajmabadi commented:
> not sure what to say. I wrote diagnostc tagger not using the system and I can show you all the comment you wrote on I should be using the tagger infrastructure.

Yes.  Your diagnostic tagger had race conditions.  The reason i said you should use the tagger infrastructure was because it would not have these issue and you could have your tagger focus on the domain specific stuff relevant to it rather than reinventing the wheel.

There is no requirement that this tagger system must be used.  But it's a *very good idea* to use it because it's battle tested and avoids these sorts of problems.

> I would prefer bunch of library/helper that one can reuse. there can be some base type like this but those bunch of library/helper will let people to create different type but share behavior for one off behavior. 

That is how tagging works.  There is a base type, and helpers around it.  Core behavior that taggers must consider in order to work well in an async environment and in order to get hte benefits of hte existing async-tagging system are properly called out.  Now, it's possible for someone to not use this system.  However, if they do, they'll need to reimplement all the logic for:

1. doing things asynchronously *in a race-free* manner.
2. Figuring out when/how they should pause to alleviate system pressure.
3. Figure out proper timing and backoff.
4. Figure out how to work across multiple buffers (i.e. when you have multiple submissions)
5. Figure out how to aggregate and dedupe tags over time.
6. Figure out how to minimally update the editor.
7. Figure out what should happen to tags when teh editor is updated.
8. Figure out how to enable/disable tagging entirely.

These are core facets that a well-behaved async tagger has, that all our tagger impls do not have to be concerned with because the existing impl handles it for them.  

This is why taggers (like the BraceMatchingTagger) can be so tiny, but so well behaved.  For example, in BraceHighlightingViewTaggerProvider all you'll see is a specification of whihc options control it, which events cause it to recompute, and then the logic to compute tags.  Everything else is taken care of for it in a consistent manner so it can't get that wrong.  

--

Note: these design approaches are a core virtue of Roslyn and a major consideration around how we do extensibility.  We know from experience that just exposing an interface or something too extensible, or too under-structured just means that everyone has to reinvent the wheel and will do things badly.  Instead, we've opted *across the board* for far more prescriptive systems where core-Roslyn handles most of the logic itself, and *only* places where there is a true need to customize behavior do we expose.

Indeed, this is the idea behind hte Analyzer/Fixer/CodeRefactoring subsystem as well.  These systems hardcode in a *lot* of behavior and knowledge.  But small extension points are given so others (including the IDE) can slot in, without being able to fully control or customize any part of the system.

CyrusNajmabadi commented:
> sam is working on the right way to handle this. which takes time.

Gotcha.  In the meantime, i would like to get this in.  It helps the IDE write *better*, more *consistent* anayzers and fixers.  If/when Sam gets a good options replacement in place then we can just update this single place to use it, and then all the features sitting on it will get the benefit.  That's one of the core reasons that it's good to have these sorts of scaffolding modules.  It means far less duplication, and a single place that can be updated to benefit lots of features.

heejaechang commented:
well. I was involved in the meeting as well. when we removed the IDE implementation and replaced it to compiler one. so not sure what to say.

sure, I guess when IDE tried to move to the new DiagnosticAnalzyer engine once we moved IDE to compiler diagnostic analyzer, I dont know what IDE forks decided (since I was not in IDE team at the time), but I do remember analyzer team talking doing this hack and that so that only IDE people can do stuff no other diagnostic analyzer can do. and that make it not be able to run in compiler. which basically tells that the IDE analyzer is not the compiler diagnostifc analyzer.

CyrusNajmabadi commented:
> well. I was involved in the meeting as well. when we removed the IDE implementation and replaced it to compiler one. so not sure what to say.

I'm very confused then.  A core *day 1* goal was to provide something suitable for the IDE.  And, congruent to that, the IDE features were then moved over onto this system.  If this system was not sufficient to meet IDE needs without "hacks", then this was never documented or communicated to IDE.  IDE just used the APIs that were delivered with the understanding that we were using them correctly.

If this was not the right way to do things, then i have no idea why this was the API that was actually delivered *for this purpose*.

> I dont know what IDE forks decided (since I was not in IDE team at the time), but I do remember analyzer team talking doing this hack

I don't know what this means.  The IDE was told that this was the right way to suitably use these APIs in order to replace the existing implementatoins the IDE had for these features.  Note that much of this work was done *by the analyzer group itself*.  i.e. many of the features were ported over by the analyzer group to demonstrate that the analyzer API was sufficient for IDE needs.

This was something i called in particular as something i was worried about. Specifically because it meant we could have significant testing holes, since the existing tests were written based on an understanding of how the existing features worked, but the ported code now used Compiler-Diagnostics, which were not guaranteed to be the same compared to the existin Semantic-Model approach we took.  

The work was done as part of hte actual *signing off* that the analyzer APIs were sufficient to meet both compiler and IDE needs.  As part of that, all APIs that hte IDE use were created and exposed to the IDE to use.  These were not stated to be hacks, nor was the IDE given alternative approaches to take to have 'non hack' codes.  Teh Analyzer system got buy-off because it met the goals set out from the beginning, including being sufficient for IDE needs.

heejaechang commented:
> Yes. Your diagnostic tagger had race conditions. The reason i said you should use the tagger infrastructure was because it would not have these issue and you could have your tagger focus on the domain specific stuff relevant to it rather than reinventing the wheel.

I dont know you removed all my implementation and made it yours and say there is a race condition. I also see there is a bug in workspace eventing which basically break state tracking and I see tagger basically workaround that by getting new snapshot Workspace.CurrentSolution and not rely on workspace events.

but regardless, it removed bunch of optimization I had there and I didn't use the infrastructure since tagger infrastructure didn't allow such information to propagate. all those gone saying too complex optimization but tagger has full of such optimization. so.. not sure what that is except being forced. 

...

like I said, having base type people can use when they feel it is helpful is fine. and I dont like inheritance is good API since it tents to make code to monolithic code.

heejaechang commented:
well. I dont know I wasn't in the meeting you are talking about. but the fact that IDE analyzer can't run in compiler without hack or special treatment just shows that it is not normal DiagnosticAnalyzer.

CyrusNajmabadi commented:
> I dont know you removed all my implementation and made it yours and say there is a race condition. 

There were literal race conditions.  i.e. shared data being accessed from several threads unsafely.  

>  I also see there is a bug in workspace eventing which basically break state tracking

I do not know what this means.  What bug in workspace eventing is there?  

> but regardless, it removed bunch of optimization I had there

What optimizations did it remove?

>  and I didn't use the infrastructure since tagger infrastructure didn't allow such information to propagate

Which information is not propagated?

> all those gone saying too complex optimization but tagger has full of such optimization. 

Yes.  That was exactly my point.  We don't want two sets of highly complex logic and optimizations for tagging.  That's just an enormous amount of technical cost that we have to understand and maintain.  So, standard IDE practice is to unify that complexity into a single area so that *all* features can benefit and so we can have a *single* implementation we can effectively maintain and update as necessary.

Complexity is very bad.  We've been in hte state before where each feature did their own thing and we had enormous amounts of bugs in each and every implementation.  That's precisely why i opened this PR.  Because there are so many bugs in so many IDE analyzers/fixers/refactorings *precisely* because each and every one of them has to reimplement the same logic, and they all do it slightly differently.  

Consolidating complex logic into a *single* location is extremely helpful for code health and code maintenance.

CyrusNajmabadi commented:
> and I dont like inheritance is good API since it tents to make code to monolithic code.

I don't know what this means.  Isn't the entire Analyzer/Fixer/Refactoring system designed as an inheritance model?  Do you have a suggestion for how i can use that system without inheritance?  Note: i could use an interface for some part of this.  But you're going to have to have a class no matter what (in order to expose the nested types that actually are the 'real' DiagnosticAnalzyer/CodeFixProvider/CodeRefactoringProvider.  So the interface would only be an extra level of indirection that doesn't seem to buy much.  

Do you have an alternative approach that would work here?  Thanks!

CyrusNajmabadi commented:
> well. I dont know I wasn't in the meeting you are talking about. but the fact that IDE analyzer can't run in compiler without hack or special treatment just shows that it is not normal DiagnosticAnalyzer.

Literally the initial design meetings we were having every week around analyzers.  From their core goals to their designs to the work done to implement them and use them across the IDE. :)

I was the person who had written the majority of hte IDE features that we wanted to move over onto analyzers.  So i was heavily involved in this from the beginning.  At no point was is a non-goal for the IDE to use this system.  It was a *primary* goal with hte intent being that hte feature needed to support this from the start.

The IDE was one of the first consumers here (outside of the simple test examples we used to vet things).  That work was done to actually prove out that hte analyzer API was sufficient.  As such, many IDE features were used to actually serve to demonstrate that this was feasible.  This included things like generate-method, simplify, add-using and others.  Unfortunately, this code all stretches back to before Roslyn became open source.

CyrusNajmabadi commented:
Are there any changes actually requested for this PR?  If not, can we merge in so it can be used to drive consistency across the IDE feature-set?

CyrusNajmabadi commented:
@heejaechang @sharwell can you let me know what, if anything, you'd like me to do here?  I would very much like this to get in as it will be the basis of many updates i'd like to make to make the IDE analyzers/fixers/refactorings operate in a consistent and understandable manner.  But i don't want to proceed on that if there are still things that need to change here.  

Please advise soon as this PR is over 6 weeks old at this point.

heejaechang commented:
as I said, I am fine it being in as long as it is used for people who want and not forced to be used for people who don't want to use this. codewise looks fine to me.

CyrusNajmabadi commented:
Ok.  Given that, can this actually get a sign-off and be merged in?  Thanks!

jinujoseph commented:
@sharwell  reviewing currently

CyrusNajmabadi commented on src/Features/Core/Portable/CodeStyle/AbstractCodeStyleProvider.cs:
```diff
@@ -0,0 +1,141 @@
+        #region analysis
+
+        protected abstract DiagnosticAnalyzerCategory GetDiagnosticAnalyzerCategory();
+        protected abstract bool DiagnosticsForOpenFileOnly(Workspace workspace);
```
Note: it seems like it would make sense to still have this in this PR.  However once #26758 goes through, it would be removed, with all affected subclasses updated accordingly.

sharwell replied:
:memo: We can just avoid migrating analyzers to the new form if they need to return `true` for this method (the one case I know about is unused usings, but that analyzer isn't a code style analyzer).

CyrusNajmabadi commented on src/Features/Core/Portable/CodeStyle/AbstractCodeStyleProvider.Refactoring.cs:
```diff
@@ -0,0 +1,89 @@
+            switch (severity)
+            {
+                case ReportDiagnostic.Suppress:
+                case ReportDiagnostic.Hidden:
```
I'm definitely a bit confused here.  My understanding of hte hierarchy was:

None: Completely disabled.  
Hidden/Silent: Enabled, but only exposed as a refactoring (i.e. a screwdriver, not as a lightbulb)
Info/Warning/Error: Enabled, exposed as a lightbulb.

sharwell replied:
Hidden severity diagnostics are *supposed* to use the screwdriver. They are refactorings that perform analysis in advance and can offer Fix All, and perhaps most importantly, can be implemented in NuGet packages.

CyrusNajmabadi replied:
>Hidden severity diagnostics are supposed to use the screwdriver. They are refactorings that perform analysis in advance and can offer Fix All, and perhaps most importantly, can be implemented in NuGet packages.

Note: *currently* i think there is no way to accomplish that.  Specifically 'refactorings' (i.e. things provided by CodeRefactoringProvider) cannot have FixAll.  After all, they have no FixAllProvider exposed on them to actually perform the fix-all.

Conversely, all CodeFixProviders (which can have fix-alls) are always shown as lightbulbs not screwdrivers.

--

Would yo ube willing to take this offline to gitter?  This area is a bit confusing, and it might be easier to hash things out 1:1 to get totaly clarity on how this all works now, what we can do right now, and where we want to get to in the future.

CyrusNajmabadi commented on src/Features/Core/Portable/CodeStyle/AbstractCodeStyleProvider.Refactoring.cs:
```diff
@@ -0,0 +1,89 @@
+
+            var computationTask = analyzerActive
+                ? ComputeOpposingRefactoringsWhenAnalyzerActiveAsync(document, span, option, cancellationToken)
+                : ComputeRefactoringsWhenAnalyzerInactiveAsync(document, span, cancellationToken);
```
I believe it's because you need hte actual option's current value to determine what the 'opposing refactorigns would be'.   Since you're gicing the *opposite* direction of whatever the analyzer is doing based on the current option.

But if the analyzer is completely inactive, you don't care what teh value is for the analyzer.  Becuase you're always offering all available refactorings that you can since the value of the option doesn't affect the refactoring itself.

CyrusNajmabadi replied:
Hard to explain in the abstract, so let's use a real world example.   Say you have "prefer block body" as "true:info".  In this case the analyzer is responsible for all the lightbulbs that offer to convert from expression to block body.  The refactoring then is only responsible for the 'opposite'.  i.e. offering the screwdriver on "convert from block to expression body".

However, say you have this set to "true:none".  Then the analyzer won't ever run since the feature is in 'refactoring mode only'.  In that case, the refactoring will light up on *both* expression and block bodies, and will offer hte screwdriver to convert to the other form.

--

This pattern is one of the reasons this PR is useful.  It codifies and enforces that.  Normally, the refactoring provider should never need to know the option value.  Because the idea of refactorings is that (as long as things aren't completely disabled) they are offered in every direction to be useful.

CyrusNajmabadi commented on src/Features/Core/Portable/CodeStyle/AbstractCodeStyleProvider.cs:
```diff
@@ -0,0 +1,141 @@
+        /// <summary>
+        /// Subclasses should implement this to provide their feature as a refactoring.  This will
+        /// be called when the user has the code style enabled, but has set it to 'refactoring
+        /// only'.
```
Oh interesting.  I guess my understanding is confused.  I think i need a full rundown on what each ReportDiagnostic and DiagnosticSeverity value means.

CyrusNajmabadi commented:
test ubuntu_16_debug_prtest

CyrusNajmabadi commented:
@sharwell Just a reminder to take this in for review.  The specific issue in question that we could not reach a resolution on was hte best way to codify internally how an "SomeOption - Refactoring Only" choice shoudl be represented internally.

My position is that we treat that as meaning: This should actually just run as a refactoring.  This has several benefits on an implementation level.  

1. it matches what hte user asked for, which i think is more intuitive for people implementing the feature (i.e. it's easy for them to say 'what should i do here?')
2. it prevents unnecessary diagnostics from being created.  i.e. since it's a truly a refactoring an implementation level, we only analyze where the caret/selection is, not every nodes/symbol/whatever that would normally be registered.
3. it ensures the UI looks exactly right and behaves exactly as if this was a refactoring.  i.e. you get the screwdriver.  The priority in the light-bulb list is appropriate based on how we currently prioritize fixes vs refactorings.  
4. it means we don't have any complexity around fix-all.  Since these are not actually fixes/diagnostics, they don't accidentally interact with the fix-all system, and there doesn't need to be any complexity in the fixers to determine what they should/shouldn't do with these.

Note: it's precisely the above sort of complexity that motivated me to make this scaffolding in the first place.  It's not only hard to get the above right, but even if you get it right, the user experience is still wrong because of issues elsewhere in the stack.

The approach you wanted (making 'Refactoring Only' options still be implemented as analyzers with diagnostics+fixes) has the only benefit of:

1. the purported claim that this is what 3rd parties do today.

However, this falls flat for me.  I'd much rather have us offer some sort of prescriptive scaffolding (like what this PR offers) that guides people into using the right portions for the right purposes to drive consistency and simplicity.  Specifically, this sort of scoffolding would help *3rd* parties as well get 1-4 correct and consistent with us.

CyrusNajmabadi commented:
@sharwell Any updates on the design questions here?

CyrusNajmabadi commented:
@sharwell Any updates on the design questions here?  Thanks!

CyrusNajmabadi commented:
@sharwell @jinujoseph  Any updates on the design questions here? This is nearly 3 months old at this point...

jinujoseph commented:
@CyrusNajmabadi sorry for the delay. We had few other priority issues to be cleared last two meetings and could not pick this. Will update by this Monday.

sharwell commented:
We discussed this in a design review today, and felt like the concepts were strong but it's not the right time to incorporate this change. During the upcoming refactoring of diagnostic analyzers from the features layer to the code style layer, we will likely want to implement logic similar to this pull request via helper methods that can be used by analyzers in the code style layer and refactorings in the features layer without coupling the layers.

For the specific question of whether Hidden severity runs through the analyzer or refactoring code path, it didn't seem to matter especially one way or the other from a user's perspective. A slight edge was given to running it in the analyzer so Fix All behavior is still allowed in the direction of the preference, and this also requires slightly fewer checks in the analyzer side (severity checks are not required for the analyzer because the behavior follows the driver default).

CyrusNajmabadi commented:
> we will likely want to implement logic similar to this pull request via helper methods that can be used by analyzers in the code style layer and refactorings in the features layer without coupling the layers.

Why close this PR out?  Isn't that what this PR is doing?  it's a common pattern for a single base type that implements the logic in a consistent fashion, but which can then be exposed by the different layers as appropriate.  

If this PR were taken, it would *easier* to perform that later architectural change, as you'd only need to update this location instead of having to redo all the features onto that new system...

Also, this PR is intended to be an internal impl detail for now.  One that can mean that our analyzers/refactorings/fixers behave in a consistent fashion.  Why would we need to wait on that?

CyrusNajmabadi commented:
> For the specific question of whether Hidden severity runs through the analyzer or refactoring code path, it didn't seem to matter especially one way or the other from a user's perspective. A slight edge was given to running it in the analyzer so Fix All behavior is still allowed in the direction of the preference

It seems very weird to have 'fix all' on something for which there is no indication in the interface at all as to what would be fixed.  It's doubly strange because we call the feature "refactoring only"...

> and this also requires slightly fewer checks in the analyzer side (severity checks are not required for the analyzer because the behavior follows the driver default).

But we do have these checks today.  And they're very necessary.  Because we often do *not* want our fixer fixing up hidden diagnostics.  Indeed, that's the default thing we do for a *ton* of fixers.  This is really important as sometimes you'll have visible and non-visible diagnostics, and when you say you want to fix the visible ones, it better not fix the non visible ones.

All this PR does is codify into a *central location* all the policy decisions we've forced onto individual features.  This is esp. important as we've seen this been done wrong across so many features, and *even now* many features don't get this right.  I'm a little surprised we would punt on this given the gaps here, even though this PR seems to effectively strictly makes things more consistent and makes it easier to do the right things in our features.  Without this, people still need to continue to write analyzers/fixers/refactorings the old way, which virtually guarantees they're going to screw up at least part of this behavior, *even if they try to be super careful*.
